spec/unit/application/device_spec.rb

```diff
@@ -318,6 +320,38 @@
       expect { @device.main }.to exit_with 1
     end
 
+    it "should exit if apply is set, and target is not" do
+      @device.options.stubs(:[]).with(:apply).returns('file.pp')
+      Puppet.expects(:err).with(regexp_matches(/missing argument.* --target/))
+      expect { @device.main }.to exit_with 1
+    end
+
+    it "should error if target is passed and the apply path is incorrect" do
+      @device.options.stubs(:[]).with(:apply).returns('file.pp')
+      @device.options.stubs(:[]).with(:target).returns('device1')
+      device_hash = {
+        "device1" => OpenStruct.new(:name => "device1", :url => "ssh://user:pass@testhost", :provider => "cisco"),
+      }
+
+      Puppet::Util::NetworkDevice::Config.expects(:devices).returns(device_hash)
+      File.expects(:file?).returns(false)
+      Puppet.expects(:err).with(regexp_matches(/does not exist, cannot apply/))
+      expect { @device.main }.to exit_with 1
+    end
+
+    it "should run an apply" do
+      @device.options.stubs(:[]).with(:apply).returns('file.pp')
+      @device.options.stubs(:[]).with(:target).returns('device1')
+      device_hash = {
+        "device1" => OpenStruct.new(:name => "device1", :url => "ssh://user:pass@testhost", :provider => "cisco"),
```

Comment:
This test failed earlier because no options are provided in "device1". Adding :options => {} makes the test succeed. Why? Well, on this line: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/network_device.rb#L8 the device.options is passed as a parameter with a nil value to Puppet::Util::NetworkDevice::Device#new() and if that parameter is passed, the default initialization to a {} performed by the #new method doesn't take place.

Reply:
As the test stands now, with the failing expectations removed, it no longer tests that apply runs (and it doesn't).